From e703de76d1d513583e871758cc91c1bdb04a2a64 Mon Sep 17 00:00:00 2001
From: Cui GaoSheng <cuigaosheng1@huawei.com>
Date: Sun, 28 Mar 2021 23:13:02 -0400
Subject: [PATCH] arm32: kaslr: Fix the bug of hidden symbols when
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: e703de76d1d513583e871758cc91c1bdb04a2a64

 decompressing code is compiled

hulk inclusion
category: bugfix
bugzilla: 47952
CVE: NA

------------------------------------------------------------------------

Fix the bug of hidden symbols when decompressing code is compiled,
we can't enable hidden cflags because decompressed code needs to
support symbol relocation.

Signed-off-by: Cui GaoSheng <cuigaosheng1@huawei.com>
Reviewed-by: Xiu Jianfeng <xiujianfeng@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/arm/boot/compressed/Makefile     | 5 +++++
 arch/arm/boot/compressed/decompress.c | 4 ----
 arch/arm/boot/compressed/misc.c       | 4 ----
 3 files changed, 5 insertions(+), 8 deletions(-)

diff --git a/arch/arm/boot/compressed/Makefile b/arch/arm/boot/compressed/Makefile
index 272046bd5a57..2d610d0a2bfb 100644
--- a/arch/arm/boot/compressed/Makefile
+++ b/arch/arm/boot/compressed/Makefile
@@ -113,6 +113,11 @@ targets       := vmlinux vmlinux.lds piggy_data piggy.o \
 
 clean-files += piggy_data lib1funcs.S ashldi3.S bswapsdi2.S hyp-stub.S
 
+ifeq ($(CONFIG_RELOCATABLE),y)
+HIDDEN_STR := -include $(srctree)/include/linux/hidden.h
+KBUILD_CFLAGS := $(subst $(HIDDEN_STR), , $(KBUILD_CFLAGS))
+endif
+
 KBUILD_CFLAGS += -DDISABLE_BRANCH_PROFILING
 
 ccflags-y := -fpic $(call cc-option,-mno-single-pic-base,) -fno-builtin \
diff --git a/arch/arm/boot/compressed/decompress.c b/arch/arm/boot/compressed/decompress.c
index 38a5dd847e03..aa075d8372ea 100644
--- a/arch/arm/boot/compressed/decompress.c
+++ b/arch/arm/boot/compressed/decompress.c
@@ -1,10 +1,6 @@
 // SPDX-License-Identifier: GPL-2.0
 #define _LINUX_STRING_H_
 
-#ifdef CONFIG_RANDOMIZE_BASE
-#pragma GCC visibility pop
-#endif
-
 #include <linux/compiler.h>	/* for inline */
 #include <linux/types.h>	/* for size_t */
 #include <linux/stddef.h>	/* for NULL */
diff --git a/arch/arm/boot/compressed/misc.c b/arch/arm/boot/compressed/misc.c
index 974a08df7c7a..abc083b0db96 100644
--- a/arch/arm/boot/compressed/misc.c
+++ b/arch/arm/boot/compressed/misc.c
@@ -16,10 +16,6 @@
  *  which should point to addresses in RAM and cleared to 0 on start.
  *  This allows for a much quicker boot time.
  */
-#ifdef CONFIG_RANDOMIZE_BASE
-#pragma GCC visibility pop
-#endif
-
 unsigned int __machine_arch_type;
 
 #include <linux/compiler.h>	/* for inline */
-- 
2.26.2

